### ISSUE TITLE
Performance degradation on TPCC 13K (and 15K)

### ISSUE THREAD

**Issue by aayushshah15** [cockroachdb/pebble#934]
We're seeing some performance degradation on TPCC 13K running on 15 `c5{d}.4xlarge` nodes compared to RocksDB. After doing the initial import, and letting the cluster rebalance with 5 partitions, and then running `./workload run tpcc --warehouses=15000 --active-warehouses=13000 --partitions=5 --ramp=1m --duration=30m {pgurl:1-15}` we see the following results.

```
# 13K TPCC on pebble
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s   104845.6  62.7%  10319.7   3758.1  27917.3  42949.7  94489.3 103079.2
```
```
 #13K TPCC on rocksdb
_elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s   161258.8  96.5%    180.7    151.0    385.9    469.8    671.1  77309.4
```

Reproduction steps:
```
set -e
set -x 

export SSDTPCC15="$USER-ssd-tpcc-storage-debug";
build/builder.sh make bin/workload;
build/builder.sh mkrelease amd64-linux-gnu;
make bin/roachprod

roachprod create $SSDTPCC15 --local-ssd=true --aws-machine-type-ssd=c5d.4xlarge --nodes=16 --clouds=aws;

echo "putting workload and cockroach binaries onto their respective nodes..."
roachprod put $SSDTPCC15:16 bin.docker_amd64/workload workload;
roachprod put $SSDTPCC15:1-15 cockroach-linux-2.6.32-gnu-amd64 ./cockroach;

echo "starting the cockroach nodes....."
roachprod start $SSDTPCC15:1-15 --args="--cache=0.25 --max-sql-memory=0.4" --racks=5

echo "configuring the cluster for fast import..."
roachprod run $SSDTPCC15:1  "./cockroach sql --insecure -e \"alter range default configure zone using num_replicas=1; SET CLUSTER SETTING schemachanger.backfiller.max_buffer_size = '5 GiB'; SET CLUSTER SETTING kv.snapshot_recovery.max_rate = '128 MiB'; SET CLUSTER SETTING kv.snapshot_rebalance.max_rate = '128 MiB';\" ";

echo "importing..."
roachprod run $SSDTPCC15:16 './workload fixtures import tpcc --warehouses=15000 --partitions=5 {pgurl:1}' 2>&1 | tee tpcc15kimport-localssd.out;

echo "upping the replication factor back to 3";
roachprod run $SSDTPCC15:1  "./cockroach sql --insecure -e \"alter range default configure zone using num_replicas=3;\"";

echo "done"
```
After this, we let the cluster up-replicate and then run:
```
roachprod run $SSDTPCC15:16 './workload run tpcc --warehouses=15000 --active-warehouses=13000 --partitions=5 --ramp=1m --duration=50m {pgurl:1-15}'
```

**Comment by aayushshah15** [comment]
cc @itsbilal for adding our updated understanding to this issue.

**Comment by itsbilal** [comment]
I took a look at the MANIFEST file out of one of @aayushshah15 's "bad" runs with Pebble. The thing that struct me the most was how the L0 read-amp was blowing up due to wide L0 files created as a result of intra-L0 compactions. That's a fairly simple fix, and is in progress (+ I'll work on backporting that to 20.2 soon, as it's a bit of a regression with rocksdb): https://github.com/cockroachdb/pebble/pull/919

After I made a binary with that change, things started to get murky. Occasionally, I saw clusters where disks on one node one node would start to slow down on writes, throwing both RocksDB and Pebble into a state where read-amp would keep climbing as compactions couldn't keep pace with flushes, and over time, TPCC query P95 latency would go up to over a minute. This is definitely a hardware/AWS issue as we see disk latency go up and throughput seesaw between 0 and intermittent spikes. Isolating and removing these nodes helps improve things very quickly, so for future runs I'd advise creating n more nodes than we need but not `cockroach start`-ing them (where n = number of partitions), unless one node goes bad and then the corresponding node among the extras can be subbed in as a clean node and the run can be restarted after that happens.

Here's an example of a "bad node" (see orange line):

<img width="554" alt="Screen Shot 2020-09-22 at 3 54 13 PM" src="https://user-images.githubusercontent.com/345332/94031642-18696d80-fd8d-11ea-9649-bd11d902a469.png">

The `cr.node.sys.host.disk.write.bytes` metric for that node would also occasionally drop down to 0 for a minute at a time, while on all other nodes it stayed consistent.

Then, I noticed that Pebble's compaction throughput at high-write-intensive points was lower than RocksDB, and that the number of compactions tended to spike and drop a lot, instead of converging at a steady state to keep pace with a steady throughput of flushes. On this front, I made two changes: 1) To decouple compaction concurrency from L0 read-amplification exclusively, and to add a second signal based on estimated pending compaction bytes; with one additional concurrency slot for every 1GB of compaction debt, and 2) To exclude flushes from the count of in-progress compactions when calculating compaction concurrency slots (this is okay as the upper bound on concurrent compactions is still respected by both flushes and compactions combined). I've made this change locally and I'm going to spin up a PR on this soon.

The third helpful change was to minimize the read/write IOPS spike that happens at the start of the TPCC workload, as load quickly ramps up. This is particularly noticeable for reads, and happens when a lot of read queries and compactions kick off at the same time. If this slows down some queries "just enough" that they back up in memory, we could potentially never recover as we're already maxing out CPU/disk capacity on queries, compactions, and flushes. The impact of compactions on this can be minimized pretty easily by jumping ahead to OS-level readahead from the start (and is a change that @aadityasondhi will finish very soon: https://github.com/cockroachdb/pebble/issues/854 ). If  I'd still encourage `--ramp`s at 3 minutes or longer for TPCC-13k to prevent this from happening, but with this change, that's a bit less of a requirement.

As part of this investigation, I also realized that Pebble periodically syncs WALs to "smooth out" writes (and to prevent the OS from causing disk latency spikes when it chose to dump a lot of buffered data onto disk), but RocksDB doesn't. For WALs in particular, this may be unnecessary as a lot of batch commits already require a WAL sync before returning. For performance parity with RocksDB I'm open to making the Pebble behaviour in line with RocksDB on this, but in my own testing I didn't notice a significant improvement in write performance with this change other than that induced by all of the above.

After all of the above, and with the occasional bad nodes isolated out (nodes that also affected RocksDB), I consistently get RocksDB-comparable performance on Pebble with tpcc-13k:

```
elapsed_______tpmC____efc__avg(ms)__p50(ms)__p90(ms)__p95(ms)__p99(ms)_pMax(ms)
 1800.0s   160420.0  96.0%    303.1    285.2    570.4    671.1    939.5   4160.7
```
